### DISCUSSION TITLE
Branching, forking, building and testing

### DISCUSSION THREAD

**Discussion by stratofax** [ion-oset/developer-commons#27] [Infrastructure]
## Ideas on how to contribute code and automate building and testing production code.

A proposed workflow for members of the OSET development team to ensure that production code committed to `main` is tested, reviewed, and free of known defects, while taking best advantage of automated testing and tracking tools offered by GitHub.

## Terms 

* **Repo** is short for a GitHub repository
* **production code** is code that's been committed to the main branch of the furthest upstream repo, the **production repo**.
* The **production repo** is the repository that's furthest upstream, and includes the code on the main branch that's intended for public release.

## Workflow overview

```mermaid
flowchart TB
  id1[Discussion: new code]
  id2[File bug report on Issues]
  id3[Create issue]
  id4[Create dev branch on your fork]
  id5[Close PR to merge fork branch]
  id6[Create PR to merge dev branch to main]
  id7[Run automation]
  id8[Close PR & Issue]
  id1 --> id3
  id2 --> id3
  id3 --> id4 
  id4 --> id5 
  id5 --> id6
  id6 --> id7 
  id7 --> id8 
  id3 --> id8
```
  
## How to merge code to the main branch of the production repo

This workflow is designed so that merging code to the main branch on the production repo will automatically:

* Trigger testing and builds
* Close a PR and a corresponding issue
* Update your repo's Projects

### Setting up Branch protection rules

Before you start using this workflow, set up Branch Protection to enforce the required policies.

To ensure all merges require a pull request, set up the following branch protection rules (under repo **Settings : Branches**).

1. Confirm Default branch is **main** (or **master** depending on your branch name)
2. Next to _Branch protection rules_, click **Add Rule**
4. Branch name pattern: **main** (or **master** depending on your branch name)
5. Protect matching branches: check **Require a pull request before merging**
6. Protect matching branches: check **Include administrators** to ensure all coders need to merge from a forked branch.
7. Click **Save Changes**

#### Other branch protection options to consider

Every box you check will potentially slow down the development process, so choose wisely, balancing the needs of protecting your code base with the added complexity of adding additional checks when you merge onto main.

Here are some additional options that may be worth the extra work.

* **Require approvals** -- just make sure your reviewers are responsive,  otherwise this can hold up progress.
* **Require review from Code Owners** -- this requires a CODEOWNERS file; see [About code owners - GitHub Docs](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) for details.

## Track fixes and new features development

With multiple contributors working on the same code, it's essential to communicate who's working on what. There are two broad types of changes to the code:

* New features
* Bug fixes

To keep things simple, we track both types of code changes in GitHub. In other words, we don't use a separate bug tracking system, and we keep discussions of new features in GitHub as well.

Tracking each type of code update is slightly different. Let's examine each separately, starting with the simpler path for tracking bug fixes.

## Capture bugs in Issues

As soon as a bug or error is discovered and reported, a repo admin or maintainer should attempt to duplicate the bug and post a detailed description to the Issues section. Then, label it as a bug.

Good bug Issues include:

* Any output or error messages.
* Information about the platform where the bug occured (Windows, Mac or Linux; iOS or Android, Firefox or Chrome). Be sure to add version numbers whenever possible.
* Version numbers for important software or dependenices (like, say, the version of Python you're using).
* Comments indicating if you've been able to reproduce the error.
* Other context or cirumstances that may contribute to the bug or error occuring.

If the assignee has any questions about the error, or requires additional information, they should post a comment to the issue, or copy relenvant posts, comments or emails from other sources so the Issue has the definitive history of all the relevant info about the error.

## Propose new features in Discussions, create a new Issue

Before adding a new feature or enhancement to the codebase, discuss the ramifications with the development team using the Discussions forum. Once the new feature is well understood, it simply requires a single click to **Create issue from discussion** --  this link is located at the bottom of the right hand column in any discussion thread.

### New featture or enhancement: Overview

1. Discuss new features in Discussions.
2. Link to any related documentation, like technical specifications.
3. Create a new Issue for the proposed code.
4. Tag the new issue as an **enhancement**.

## Creating new Issues

Whether you've discovered an error or proposed an enhancement, you'll create an Issue to track the progress of writing, testing, and merging the code.

When you create an Issue, try to connect it to:

* Assignees: who will write and test the code?	
* Labels: your new issue should include either a `bug` or `enhancement` label -- but never both! Other labels are optional but appreciated.
* Projects: the main project for the repo. By linking your Issue to the repo's Project, GitHub will automatically update the status of the issue in the Project as you create a PR, request reviews, and merge the code into `main`.
* Development: don't add anything here until you've created a pull request to merge your code into `main`, typically later in the development workflow.

Add as many comments as you like to your new Issue as you and the development team work on it. Avoid using other discussion tools (like Slack or email) to discuss the issue, unless you add a summary of these outside discussions to the Issue afterwards. The goal is to be able to follow all decisions related to the issue in one place, the Issue itself.

## Create your fork and branch

By the time code is submitted, via a pull request, to be merged onto `main`, it should be well tested and fully reviewed. 

To enable this workflow, you'll need to do your work on a fork of the production repo. This is where you'll do your work.

By forking the production repo, you create an independent sandbox where you can write and test code to ensure it's ready to be merged into the production repo using a Pull Request.

## Create an Issue and corresponding branch

All development should take place on a branch off of `main`, and this branch should be created first on the production repo when the corresponding issue is created.

1. After an issue is created, add a new branch on the production repo to track the changes for this issue.
2. Link the branch to the Issue.

## Create a branch on your fork

1. Using the _Branch source_ option, create a branch with the same name on your fork of the repo by choosing the upstream (production) repo and the branch you just created.
4. Checkout the branch on your development workstation.
5. Commit changes to the new branch on your fork.

 ## Create a pull request on the branch you checked out from your fork

This will merge the code to the branch with the same name on the production repo.

1. When you've finished committing your code, go to your fork on GitHub and select the working branch you checked out.
2. Select the commits you just pushed and create a new pull request ("This branch is  x commits ahead...").
3. Your new pull request should compare changes to the branch with the same name on the production repo.
4. Create your pull request.
5. Add reviewers to your pull request, assign it (often, to you), give it a label.
6. Typically, you do **not** want to assign your PR to a milestone.
7. When the PR is reviewed and tested, merge it with the same branch on the production repo.

Now, your new code is merged into the production repo, on the same branch name that you checked out on your fork.

## Clean up your fork

If your recent Pull Request resolves the issue, you can delete the branch on your fork. GitHub will display the **Delete Branch** button after the PR is merged and closed. 

1. If you're done with the changes for the issue, you can delete the branch from your fork.
2. Also, be sure to delete the checked out branch from your development workstation (or IDE).

## Merge onto main

If you think your code is ready to merge onto the `main` branch on the production repo, create a PR.

### Create a PR from the working branch

1. Go to the working branch you created on the production repo (click **Code** and select your working branch).
3. Select "**x commits ahead** on this branch compared to main."

This should create a pull request that targets the main branch from your working branch.

**Tip:** Always review the line directly below the Pull request title to confirm that your pull request is pulling changes from and to the correct branches.

### Create a PR from the Pull Requests tab

1. Click on **Pull requests** (between Issues and Discussions).
2. Click on the **New pull request button** (top right)
3. Select the target and source branches for your pull request.

### Create the Pull Request

5. Give your PR a descriptive title.
6. Add a detailed description.
7. Review the commits, files changed, and Create a pull request.
8. While you're waiting for the checks to complete, add reviewers, assign the PR, add labels, select a project, and **select an Issue from Development**. This is the best time to link your PR to one or more open issuse. 
9. When your reviewers approve your changes, merge your PR from the command line:

```
git checkout main
git merge development
git push
```

These commands will close your PR on GitHub with a fast-forward merge, without creating an additional commit hash.

**Close issues when your pull request merges onto main**. 

If your PR completes the work on the Issue, or a milestone, delete the working branch from the production repo if the branch is intended to track a new feature or resolve a bug

Note that some branches may persist for many issues.  

### Sync your fork

Go to your fork on GitHub. 

On the main branch of your fork, you'll see you're now several commits behind the main branch of your production repo. Use the **Sync fork** link to reveal and click the **Update branch** button. Repeat this process after you switch to your development branch.

Both your main and development branch on your fork should be synced with the corresponding branches on your production repo.

Now, there's one last step.

### Sync your development workstation

Open up your development directory or favorite IDE or code editor and sync your development branch on your workstation so you have a local copy of all the updates you've just made to your fork. 

## Fork and branch workflow overview

1. All code contributors, including the project owners, must create a fork of the project to their GitHub account.
12. Then, create and check out a branch on your fork to develop and test your code. This fork should correspond to an open Issue on the production repo. 
13. Add a comment on the correspoding Issue on the production repo that you've create a branch on your fork to work on the issue. Be sure to include a link to your new branch.
14. When you've tested your code and want reviewers to read and run your new code, create a PR on the production repo.
15. Reviewers can compare code and sync to your fork if they want to run your code locally.
16. Owners may merge the fork on your branch to a matching branch on the production repo. 

### Testing on your fork

It's essential that you write tests on the working branch on your fork, especially if you're adding a new feature. If you're fixing a bug, you should verify that your work fixes an existing test, or that you've written a test that demonstrates the bug on the existing code. Your new code should pass this test.

Also, it's essential that you ensure that your new code doesn't break any existing tests, either.

**Comment by cwulfman** [comment]
Who is the intended audience? The language suggests it is developers in general, not specifically contributors to TTV projects.

**Reply by stratofax** [reply]
Intended audience now identified in the first graf.

**Comment by stratofax** [comment]
A question about the "Merge onto main" step: which approach is best?

- Merge commit (default)
- Squash and merge
- Rebase and merge

I've been using the default **Merge commit** but I'm thinking one of the other two options will help keep the main branch merge history as simple and easy-to-follow as possible.

**Reply by stratofax** [reply]
Based on recent discussions, the "merge commit" is preferred for persistent branches; squash and rebase are good options for short-lived branches.

**Reply by ion-oset** [reply]
From my notes...

For merges into upstream:

- _Squash_ for branches with small numbers of changes that are essentially one big commit. The thing to check is that no information is lost if you compare the squash to the original branch. For example if there were subtle bugs that were cleaned up, and no other note of them was made, the history is the only record of a path that should not be taken again.
- _Rebase_ for complex branches where the history captures important decisions, or where a change was complex and allowance should be made for rolling it back partway.
- _Merge_ for long-lived branches (e.g. ongoing development of a milestone, or release branches), or when there have been other changes on the upstream branch that would be complex to do a rebase on,

Development branches from forks should be merging first into the staging branch on upstream (whatever it may be called), and not directly into the production branch (by our convention: `main`). If the merge from development to staging is effectively a fast-forward (there's been no change to the staging branch), then use _rebase_ instead of _merge_ instead (to avoid a useless merge commit in GitHub).

**Reply by stratofax** [reply]
These are great notes. @ion-oset do you want to incorporate them into the doc as you move it to the code tree? Alternatively, I can add it in later, when you've finished the transfer.

**Comment by stratofax** [comment]
@cwulfman @ion-oset I added some minor edits and cleaned up the last few sections with some strategic deletions. @ion-oset will add this doc to the source tree at some point -- let us know when that's happening, so I don't edit this version after it's been transferred.

**Reply by stratofax** [reply]
I also updated the flow diagram

**Comment by stratofax** [comment]
Here's a rule of thumb that I think is important and deserves emphasis:

**Only committing to main should close Issue.**

I actually say this in the doc (above) in the **Merge onto main** section (step 6) but I think it's worth saying: **don't** close an issue if you're just merging code from your fork to the corresponding branch on the production repo.

**Comment by stratofax** [comment]
Also, I think it's important to discuss how to handle syncing your fork with the production repo after merges happen.

**Comment by windoverwater** [comment]
Uh, coming late to this thread.  Is it too late to reconsider the above?  It is probably best to discuss this in person.  However, as my fingers are still typing, the reason I mention it here/now is that the plan to have code get into master via two PR cycles seems like 2x the overhead.  Every situation is different and I am NOT plugged into the current / desired dev processes, but from a newbie POV it is just easier and less overhead to skip the fork model for an 'integrated' team; that is, a team within a single enterprise/organization.  I am thinking TrustTheVote is such a situation - yes/no?  The fork model is fine for someone outside the team as they then cannot 'pollute' the repo with bad branch names etc.  But the four us should be able to agree on a simple and consistent branch naming and PR convention.

So my suggestion is to adopt a more streamlined no-fork model for ElectOS-Dev team members who have Write permission (for good reason) and use a fork model for external ElectOS-Dev contributors who do not have Write permission (for good reason).

Then for us there is no need to solve the syncing your fork after merge issue.

?

**Comment by cwulfman** [comment]
I'm inclined to agree with you, @windoverwater .  The fork-and-merge strategy does seem like overkill now that we have teams.

**Comment by ion-oset** [comment]
It's not too late to reconsider. We are trying this out to see how well it works, but this is a proposal. Consider it an RFC in the draft stage. Nothing is firmly committed to and details are up for discussion. But we do have reasons for wanting to do it this way.  Let me address what I think are your main points @windoverwater and explain at least my view of it.

**Reply by ion-oset** [reply]
**Do we need two levels of review?**

The proposal has two levels for a couple of reasons:

1. To model the common stages of deployment: development, staging and production. Those stages don't have to map to branches but it's convenient if they do. What those branches are named and which repositories they are on depends on situation.
        
    For our purposes I think it's important that production specifically be the `main` branch on the upstream (TrustTheVote-Project owned) repository. The reason is that if outsider clone the repository they will almost always go straight to `main` to try it out. That branch should be as stable as possible. But nothing prevents work from ending up on whatever the staging branch is for some time.

2. GitHub Actions run when a PR is merged. We want a place to test the merged version before it lands on `main` to minimize chances of `main` breaking.

**Reply by ion-oset** [reply]
**Do we need to use forks instead of just a lot of branches in the same repository?**
  
My experience is that forks have a lot of advantages:
    
- More flexible. They allow a proliferation of experiments, most of which aren't ready to share or see the light of day. If more than one person is working against a branch, there's no question of whether it's OK to do something with it.
- Less brittle. If the repository gets into an inconsistent state it's much more likely to be on a fork, and less likely to break things for a team.
- Facilitating ongoing review of complex work. The use of Draft PRs to get in-progress review can be done in the fork. If the PR needs a serious re-do in place or gets dropped that doesn't get tangled into the upstream project. No one gets stuck with inconsistent branches if they cloned upstream. Force pushes can be forbidden on upstream, but are fine on one's own fork.
- Plays nicer with automation. Branches on forks aren't restricted by whatever policies branches on upstream are following until they are ready to merge.
- Easier to navigate. A repository full of partial branches gets confusing. Forks allow us to be stricter about what's in shared space, and means a new clone doesn't get a lot of branches being used by other people.
- Less chance of losing work to local failures. If one is in the middle of work that's not ready for review, pushing to the fork means it's backed up even if it's broken.
- We can track our own tasks as specific issues without cluttering the shared space.
- We still want a consistent naming convention, but we don't have to get it all worked out in advance.

**Reply by windoverwater** [reply]
Mmm, the above references a number of pros and cons.  And though I agree with most of the points, for me it is the 'value' or importance/impact of each specific item/choice as summed over the final solution that represents the better choice(s).  For example, for #2 of the first comment, I would value low since GitHub actions can run on just about any GitHub event at this point.  And tests etc need to run prior to someone looking at a PR, not just after merging.  (Running them prior to looking at a PR usually saves more time.)

One of my favorite comedies (this will date me) is 'My Cousin Vinny'.  One reason is in the movie the protagonist selects the strategy of 'looking at every brick in wall' and seeing if it is individually valid/good.  

Regardless, I have not yet used GitHub workflows/actions as I have always been stuck with Jenkins or something else or something less.  So I am on thin ice here!  And as such I am not sure it worthwhile for me at this time to second guess any of the above given that it is already written down post thought.  It can be changed later if necessary.

The best I can do is probably explain some general rules of thumb that I use which is what is causing me pause regarding the double PR process above:

- no process before it time
- optimize for the good (efficiency) and not for the bad (taxation)
- protect against the very bad

If this is important perhaps we can talk about it next Tuesday in person on zoom?

p.s. note - my original comment was intended to ask the question if we could support both, fork and intra repo branching, not just one or the other:  intra-repo branching for intra-team efficiency and fork-based whenever the extra step/isolation is desired or necessary

**Comment by stratofax** [comment]
Because this process isn't complex enough, I added a new section near the bottom, Post merge sync-up. I think a squash or rebase merge might make sense here. I'll test it out next time around.

**Comment by stratofax** [comment]
I saw this error twice yesterday when running my GitHub Actions. It seems like a GitHub problem, because when I run the same action again, it works without modification.

```
Updating dependencies
Resolving dependencies...

unexpected http resp 429 for https://github.com/ion-oset/nist-data-models.git/info/refs?service=git-upload-pack
Error: Process completed with exit code 1.
```

Notably, this is related to the repo reference in BallotLab's `pyproject.toml` file.

**Reply by ion-oset** [reply]
@stratofax I have a suspicion as to a possible cause. But I think this should go into a BallotLab issue or discussion, since this thread has been focused on a general process not debugging. Are you OK with moving it? I'll post my theory wherever you choose.

**Reply by stratofax** [reply]
Yes, I agree, @ion-oset, this is more suitable for a BallotLab discussion. I don't think it should be an Issue, because it isn't easy to replicate, and doesn't seem directly related to code execution.

I didn't actually see the connection to nist-data-models.git until I started writing the original post, and kept clicking through the error messages until I found the one I posted (above), which in fact included some useful info. After I posted, I was like, "this is probably better tracked on the BallotLab repo."

So yes, please post your theory on the BallotLab Discussions board.

**Comment by ion-oset** [comment]
@stratofax said on [BallotLab #119](https://github.com/TrustTheVote-Project/BallotLab/pull/119#issuecomment-1259713160):

> I agree with the premise of the development branch, I'm proposing one option to implement this -- instead of having a long running development branch, create branches for each milestone, merge them onto main when the milestone is completed, delete the milestone branch, rinse, repeat.
>
> You're proposing another option -- maintain the development branch as a persistent target for work around a specific milestone. When we complete a milestone, use the git branch --force command to ensure that the development branch is at exactly the same point as main.

Just as a note, the persistent branch isn't my proposal. We came up with it collectively on the Sept 6th call.

Let's call them "primary development branch" and "development branch per milestone".

> Both strategies share essential features:
>
> - There's a primary development branch where all the work for a milestone lands before it's merged to `main`
> - When we reach a milestone and merge to main we ensure that the development branch and main are identical.
>
> The difference is (as far as I can tell) exactly **how** we ensure that `main` and the milestone development branch are identical immedately after merging. You suggest using a git command on a long-lived development branch; I suggest creating a new branch off of main for the upcoming milestone.

I'm OK with either approach. For the record the pros and cons I see:

_primary development branch_:

- _pro_: Everyone knows what the sync point and base point is. There's no chance of confusion.
- _pro_: Familiarity. The approach of a single branch named `develop` was made popular by the [Git Flow](https://nvie.com/posts/a-successful-git-branching-model/) branching model. The pros and cons and of that model are complex enough to be their own discussion topic, but for our current purposes it means "develop" is a known name.
- _con_: Need additional policy if multiple concurrent development targets need to land at the same time.

_per milestone development branch_:

- _pro_: More than one milestone could be run concurrently. I think this shouldn't happen very often but it can under circumstances such as just happened where a milestone needs to be re-opened.
- _con_: Tracking on which milestone to land on could be confusing.
